File: Tutorial3SuperClassEntities/src/main/java/com/example/tutorial3superclassentities/data/source/local/MeasurementDatabase.java
package com.example.tutorial3superclassentities.data.source.local;

import android.arch.persistence.room.Database;
import android.arch.persistence.room.Room;
import android.arch.persistence.room.RoomDatabase;
import android.content.Context;

import com.example.tutorial3superclassentities.data.source.model.MeasurementAnglemeter;
import com.example.tutorial3superclassentities.data.source.model.MeasurementBubbleLevel;
import com.example.tutorial3superclassentities.data.source.model.MeasurementLaser;
import com.example.tutorial3superclassentities.data.source.model.MeasurementProtractor;
import com.example.tutorial3superclassentities.data.source.model.MeasurementRuler;

import static com.example.tutorial3superclassentities.data.source.local.MeasurementDatabase.DATABASE_VERSION;

@Database(entities = {MeasurementAnglemeter.class,
        MeasurementBubbleLevel.class,
        MeasurementLaser.class,
        MeasurementProtractor.class,
        MeasurementRuler.class}
        , version = DATABASE_VERSION)
public abstract class MeasurementDatabase extends RoomDatabase {

    public static final String DATABASE_NAME = "clinometer.db";
    public static final int DATABASE_VERSION = 1;

    private static MeasurementDatabase INSTANCE;

    private static final Object sLock = new Object();

    public abstract MeasurementAnglemeterDao measurementAnglemeterDao();

    public abstract MeasurementBubbleLevelDao measurementBubbleLevelDao();

    public static MeasurementDatabase getInstance(Context context) {
        synchronized (sLock) {
            if (INSTANCE == null) {
                INSTANCE = Room.databaseBuilder(context.getApplicationContext(),
                        MeasurementDatabase.class, DATABASE_NAME)
                        .allowMainThreadQueries()
                        .build();
            }
            return INSTANCE;
        }
    }
}

File: Tutorial3-1SuperClassEntities/src/main/java/com/example/tutorial3superclassentities/data/source/local/MeasurementDatabase.kt
package com.example.tutorial3superclassentities.data.source.local

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import com.example.tutorial3superclassentities.data.source.local.MeasurementDatabase
import com.example.tutorial3superclassentities.data.source.model.*

@Database(
    entities = [MeasurementAnglemeter::class,
        MeasurementBubbleLevel::class,
        MeasurementLaser::class,
        MeasurementProtractor::class,
        MeasurementRuler::class],
    version = MeasurementDatabase.DATABASE_VERSION
)
abstract class MeasurementDatabase : RoomDatabase() {

    abstract fun measurementAnglemeterDao(): MeasurementAnglemeterDao?
    abstract fun measurementBubbleLevelDao(): MeasurementBubbleLevelDao?

    companion object {

        const val DATABASE_NAME = "database_sample.db"
        const val DATABASE_VERSION = 1

        private var INSTANCE: MeasurementDatabase? = null
        private val sLock = Any()

        fun getInstance(context: Context): MeasurementDatabase? {

            synchronized(sLock) {
                if (INSTANCE == null) {
                    INSTANCE =
                        Room.databaseBuilder(
                            context.applicationContext,
                            MeasurementDatabase::class.java,
                            DATABASE_NAME
                        )
                            .allowMainThreadQueries()
                            .build()
                }
                return INSTANCE
            }
        }
    }
}